blog: introducing suspended jobs (#27197)
kubernetes/website: content/en/blog/_posts/2021-03-18-introducing-suspended-jobs.md

---
title: "Introducing Suspended Jobs"
date: 2021-03-18
slug: introducing-suspended-jobs

> url: /blog/2021/03/introducing-suspended-jobs

Reviewer: I recommend omitting the URL and letting the site generator work out the URL automatically.

---

**Author:** Adhityaa Chandrasekar (Google)

> Jobs are a crucial part of Kubernetes API. While other controllers such as

Reviewer: ```suggestion
Jobs are a crucial part of Kubernetes API. While other kinds of workload resource, such as
```

(to cluster users the resource is what matters; the fact that it's a controller doing the work and not, eg, lots of fast-typing elves is an implementation detail :wink:)

> [Deployments](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/),

Reviewer 2: remove the `https://kubernetes.io/` part

[ReplicaSets](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/),
[StatefulSets](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/), and
[DaemonSets](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/)
solve use-cases that require Pods to run forever, Jobs are useful when Pods need

> to terminate. Commonly used in parallel batch processing, Jobs can be used in a

Reviewer 2: run to completion

maintenance to sending bulk emails and scientific computing.

While the amount of parallelism and the conditions for Job completion are
configurable, the Kubernetes API lacked the ability to suspend and resume Jobs.
This is often desired when cluster resources are limited and a higher priority
Job needs to execute in the place of another Job. Deleting the lower priority
Job is a poor workaround as Pod completion history and other metrics associated
with the Job will be lost.

With the recent Kubernetes 1.21 release, you will be able to suspend a Job by

> updating its Job spec. The feature is currently in **alpha** and requires you to

Reviewer 3: its ~~Job~~

> Job by updating its Job spec. This feature is in **alpha** in 1.21 and requires

Reviewer: I would phrase this as if v1.21 is already released.

on the [API server](/docs/reference/command-line-tools-reference/kube-apiserver/)
and the [controller manager](/docs/reference/command-line-tools-reference/kube-controller-manager/)
in order to use it.

## API changes

A new boolean field `suspend` is introduced in the Job spec API. Let's say I

> create the following Job.

Reviewer: nit:
```suggestion
create the following Job:
```

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: my-job
spec:
  suspend: true
  parallelism: 2
  completions: 10
  template:
    spec:
      containers:
      - name: my-container
        image: busybox
        command: ["sleep", "5"]
      restartPolicy: Never
```

Jobs are not suspended by default, so I need to explicitly set the `suspend`

> field to true in the above Job spec. The Job controller will now refrain from

Reviewer 3: in the above example

creating Pods until I'm ready to start the Job, which I can do by updating the
field to false.

> If I update this field from false back to true again, the Job controller will

Reviewer 3: The above paragraph says you set it to true, so you should continue your thought.

Author: I've moved some words around to make things a bit less confusing, PTAL, thanks!

terminate all active Pods it created. This is done by sending the Pods a SIGTERM
signal; the [graceful termination period](/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination)
defined in the Pod spec will be honoured and Pods terminated this way will not

> be counted as failures by the Job controller.

Reviewer 4: I would explicitly indicate that success and failed pods will continue to exist, and that upon unsuspension they will count against job completion. This will set the stage for the next section discussing the where this is useful.

Author: fair point, I've added a paragraph. PTAL, thanks.

be counted as failures by the Job controller.

Read the [documentation](/docs/concepts/workloads/controllers/job#suspending-a-job)
for a full overview of the implications of using this new feature.

## Where is this useful?

Let's say I'm the operator of a large cluster. I have many users submitting Jobs

> to the cluster, but not all Jobs are created equal — some Jobs are more

Reviewer 2: treated?

Reviewer: “created” seems OK to me.

Author: "created" was intentional. It was a subtle reference to https://en.wikipedia.org/wiki/All_men_are_created_equal :)

must share resources. If all Jobs were created in the suspended state and placed
in a pending queue, I can achieve priority-based Job scheduling by resuming Jobs
in the right order.

> As another motivational use-case, consider a cloud provider where compute

Reviewer 4: we need to be careful with the why this use case is described because we count completions from scratch every time the job is unsuspended. I would suggest not to include this use case or clearly callout that the application needs to have the ability to understand how to continue after a suspend.

Reviewer 2: finalized pods are not removed (unless they are GCed). But pod progress is lost.

Reviewer: Because these shortcomings could be surprising, I think we _should_ mention a case where they might be relevant, and then explain the problem. Maybe mention Indexed completion as a way for Pods to work out what progress to (separately) track.

Author: >we count completions from scratch every time the job is unsuspended

Perhaps I've misunderstood, but we do NOT count from scratch. Successfully completed pods in the past will continue to count towards the completions counter even if the Job is suspended and resumed. Pods that were *removed* as a direct result of the suspension won't be counted, of course, but I thought that is fairly obvious from the note about SIGTERM in the previous section.

Reviewer 4: Not directly obvious, I left a comment on the paragraph above.

> resources are cheaper in the night than in the morning. If I have a long-running

Reviewer 3: cheaper ~~in~~ at night?

Job that takes multiple days to complete, being able to suspend the Job in the

> morning and then resume it in the evening every day can reduce costs.

Reviewer: Aside: a similar use case is _load shedding_ - responding to changes in the wholesale electricity price by cutting compute use (and, presumably, setting nodes to idle). Firms can earn money by agreeing to shed load in response to a signal, which might be a grid frequency shift below a threshold or a separate notification.

Since this field is a part of the Job spec, CronJobs automatically get this
feature for free too.

## References and next steps

If you're interested in a deeper dive into the rationale behind this feature and
the decisions we have taken, consider reading the [enhancement proposal](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2232-suspend-jobs).

> The official documentation is available in the [Jobs API page](/docs/concepts/workloads/controllers/job#suspending-a-job).

Reviewer: I'd write:
```suggestion
There's more detail on suspending and resuming jobs in the documentation for [Job](/docs/concepts/workloads/controllers/job#suspending-a-job).
```

The official documentation is available in the [Jobs API page](/docs/concepts/workloads/controllers/job#suspending-a-job).

As previously mentioned, this feature is currently in alpha and is available
only if you explicitly opt-in through the `SuspendJob` feature gate. If this is
a feature you're interested in, please consider testing suspended Jobs in your

> cluster and providing feedback. The SIG Apps community [meets regularly](https://github.com/kubernetes/community/tree/master/sig-apps#meetings)

Reviewer 2: maybe a link to the enhancement issue?

and can be reached through [Slack or the mailing list](https://github.com/kubernetes/community/tree/master/sig-apps#contact).
Barring any changes to the API, we intend to graduate the feature to beta in

> 1.22 and the feature will be available by default.

Reviewer: ```suggestion
Kubernetes 1.22, so that the feature becomes available by default.
```
